ipa-core/src/protocol/context/mod.rs

```diff
@@ -312,7 +316,35 @@ impl<'a, B: ShardBinding> Context for Base<'a, B> {
 
 /// Context for MPC circuits that can operate on multiple shards. Provides access to shard information
 /// via [`ShardConfiguration`] trait.
-pub trait ShardedContext: Context + ShardConfiguration {}
+pub trait ShardedContext: Context + ShardConfiguration {
+    /// Requests data to be received from all shards that are registered in the system.
+    /// Shards that don't have any data to send, must explicitly open and close the send channel.
+    fn recv_from_shards<M: Message>(
+        &self,
+    ) -> impl Stream<Item = (ShardIndex, Result<M, crate::error::Error>)> + Send {
+        stream::select_all(
+            self.peer_shards()
+                .map(|origin| self.shard_recv_channel(origin).map(move |v| (origin, v))),
+        )
+    }
+
+    /// Picks a shard according to the value obtained from sampling PRSS shared with the given helper.
+    fn pick_shard(&self, record_id: RecordId, direction: Direction) -> ShardIndex {
+        // FIXME: update PRSS trait to compute only left or right part
+        let (l, r): (u128, u128) = self.prss().generate(record_id);
+        let shard_index = u32::try_from(
```

Comment:
we discussed this below, to summarize - it shouldn't be a security issue because all shards run MPC circuits independent from each other. Each circuit runs an independent Diffie-Hellman protocol between each pair of helpers, so the probability of them negotiating the same shared secret is $\frac{N}{2^{256}}$ where $N$ is the total number of shards